Sources/Sentry/SentryHub.m

```diff
@@ -590,37 +591,64 @@ - (void)captureEnvelope:(SentryEnvelope *)envelope
 
 - (SentryEnvelope *)updateSessionState:(SentryEnvelope *)envelope
 {
-    if ([self envelopeContainsEventWithErrorOrHigher:envelope.items]) {
+    BOOL handled = YES;
+    if ([self envelopeContainsEventWithErrorOrHigher:envelope.items wasHandled:&handled]) {
         SentrySession *currentSession = [self incrementSessionErrors];
 
         if (currentSession != nil) {
+            if (!handled) {
+                [currentSession endSessionCrashedWithTimestamp:[_currentDateProvider date]];
+
+                _session = [[SentrySession alloc] initWithReleaseName:_client.options.releaseName];
+                _session.environment = _client.options.environment;
+                [self.scope applyToSession:_session];
+            }
+
             // Create a new envelope with the session update
             NSMutableArray<SentryEnvelopeItem *> *itemsToSend =
                 [[NSMutableArray alloc] initWithArray:envelope.items];
             [itemsToSend addObject:[[SentryEnvelopeItem alloc] initWithSession:currentSession]];
-
             return [[SentryEnvelope alloc] initWithHeader:envelope.header items:itemsToSend];
         }
     }
-
     return envelope;
 }
 
 - (BOOL)envelopeContainsEventWithErrorOrHigher:(NSArray<SentryEnvelopeItem *> *)items
+                                    wasHandled:(BOOL *)handled;
 {
     for (SentryEnvelopeItem *item in items) {
         if ([item.header.type isEqualToString:SentryEnvelopeItemTypeEvent]) {
             // If there is no level the default is error
-            SentryLevel level = [SentrySerialization levelFromData:item.data];
+            NSDictionary *eventJson = [SentrySerialization eventEnvelopeItemJson:item.data];
+            if (eventJson == nil) {
+                return NO;
```

Comment:
`m`: Shouldn't we also set the handled to the default of `YES` here?

Reply:
I dont think it matters, if we return `NO` there is no error, I dont think we should change the input.
This is how API with `NSError**` as parameters work.